### PR TITLE
Changes for Windows E2E tests

### PR DESCRIPTION
**Pull request by philipktlin**
1. Changes Publish-Branch.ps1 to have similar logic as what Linux does.  Only build IoTEdgeQuickstart and Leaf device with runtime (-r) parameter; in order to generate executable.
2. Update regex pattern to handle management uri with/without double quote.
3. Add retry logic for calling iotedge list if hyper error found.
4. Add logging for Edge Hub certificates.
5. Publish temperature filter function image.
6. Add new logic in LeafDevice to handle certificate validation when running on Windows.
7. Output win-x86 and win-x64 for IotEdgeQuickStart and LeafDevice.

### CODE REVIEW

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs [hunk 1]
```diff
@@ -52,24 +54,61 @@ public async Task VerifyModuleIsRunning(string name)
...
+                            else
+                            {
+                                await Task.Delay(TimeSpan.FromSeconds(5), cts.Token);
+                            }
+                        }
+                        catch (Exception listOperationException)
+                        {
+                            // Display error and retry for some transient exceptions such as hyper error
+                            string exceptionDetails = listOperationException.ToString();
+                            Console.WriteLine(exceptionDetails);
```
Comment (varunpuranik) [inline smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: WriteToConsole?

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs [hunk 2]
```diff
@@ -52,24 +54,61 @@ public async Task VerifyModuleIsRunning(string name)
...
+                            {
+                                throw;
+                            }
+
+                            if (exceptionDetails.Contains("Could not list modules", StringComparison.OrdinalIgnoreCase))
+                            {
+                                try
+                                {
+                                    Console.WriteLine("Workaround: restart iotedge service.");
+                                    await this.Restart().ConfigureAwait(true);
```
Comment (varunpuranik) [inline smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Is this ConfigureAwait(true) needed? I would avoid it if not..

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs [hunk 3]
```diff
@@ -52,24 +54,61 @@ public async Task VerifyModuleIsRunning(string name)
...
+                            WriteToConsole("Output of iotedge list", result);
+
+                            status = result
+                                .Where(ln => ln.Split(null as char[], StringSplitOptions.RemoveEmptyEntries).First() == name)
+                                .DefaultIfEmpty("name status")
+                                .Single()
+                                .Split(null as char[], StringSplitOptions.RemoveEmptyEntries)
+                                .ElementAt(1); // second column is STATUS
+
+                            if (status == "running")
```
Comment (varunpuranik) [inline smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Suggest using Edge.Util.Retry

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs [hunk 4]
```diff
@@ -137,23 +174,33 @@ public async Task Configure(string connectionString, string image, string hostna
                 string[] result = await Process.RunAsync("powershell", args, cts.Token);
                 WriteToConsole("Output from Configure iotedge windows service", result);
 
-                UpdateConfigYamlFile(runtimeLogLevel);
+                // Stop service and update config file
+                await this.Stop().ConfigureAwait(false);
```
Comment (varunpuranik) [inline smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Again, avoid the COnfigureAwait..

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs [hunk 5]
```diff
@@ -137,23 +174,33 @@ public async Task Configure(string connectionString, string image, string hostna
...
+
+                UpdateConfigYamlFile(deviceCaCert, deviceCaPk, deviceCaCerts, runtimeLogLevel);
 
                 // Explicitly set IOTEDGE_HOST environment variable to current process
                 SetEnvironmentVariable();
             }
         }
 
-        static void UpdateConfigYamlFile(LogLevel runtimeLogLevel)
+        static void UpdateConfigYamlFile(string deviceCaCert, string deviceCaPk, string deviceCaCerts, LogLevel runtimeLogLevel)
```
Comment (varunpuranik) [inline smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Nit: rename deviceCaCerts to ownerCert or TrustBundleCerts or something like that..

File: smoke/LeafDevice/details/Details.cs [hunk 6]
```diff
@@ -50,48 +56,62 @@ bool useWebSockets
...
         protected async Task ConnectToEdgeAndSendData()
         {
-            Microsoft.Azure.Devices.IotHubConnectionStringBuilder builder = Microsoft.Azure.Devices.IotHubConnectionStringBuilder.Create(this.iothubConnectionString);
+            IotHubConnectionStringBuilder builder = IotHubConnectionStringBuilder.Create(this.iothubConnectionString);
             string leafDeviceConnectionString = $"HostName={builder.HostName};DeviceId={this.deviceId};SharedAccessKey={this.context.Device.Authentication.SymmetricKey.PrimaryKey};GatewayHostName={this.edgeHostName}";
-            
-            this.context.DeviceClientInstance = Option.Some(DeviceClient.CreateFromConnectionString(leafDeviceConnectionString, this.deviceClientTransportType));
+
+            // Need to use CustomCertificateValidator since we can't automate to install certificate on Windows
+            // For details, refer to InstallCaCertificate method
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: This logic in 2 places is confusing..
I would say refactor InstallCaCertificate to something like InitializeServerCerts and in that have the if windows else logic..

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs (hunk 4 above)
Comment (philipktlin) [reply smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Updated in another PR.

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs (hunk 3 above)
Comment (philipktlin) [reply smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Updated in another PR.

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs (hunk 2 above)
Comment (philipktlin) [reply smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Updated in another PR.

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs (hunk 1 above)
Comment (philipktlin) [reply smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Updated in another PR.

File: smoke/IotEdgeQuickstart/details/IotedgedWindows.cs (hunk 5 above)
Comment (philipktlin) [reply smoke/IotEdgeQuickstart/details/IotedgedWindows.cs]: Updated in another PR.

File: smoke/LeafDevice/details/Details.cs (hunk 6 above)
Comment (philipktlin) [reply smoke/LeafDevice/details/Details.cs]: Updated in another PR.

**Pull request merged**